### PR TITLE
Add Fetch PSPs Externally + new tests + Increase the code robustness.

### PR DESCRIPTION
**Pull request by Ethnical**
**Description**

This PR is the latest PR of the milestone 1 for AP. 
This PR allow the psp-executor service to fetch the PSPs externally and execute them on chain. 
For now, only the RPC url that contains `fork` are compliant with the code. 

I have updated the documentation accordingly of with the new features added. 
I have also refactor and change the multiples variables, to increase the robusteness of the codebase. 
**CLI arguments** 
I also added new CLI arguments: 
- `--path`: The path to the JSON file with the PSPs.
- `--safe.address`: The safe address you want to fire PSPs from. 
- `--chainid`: the chainId of the network. 
and I have remove the `--receiveraddress` and `HexString` because this one will be fetch from the PSPs now. 
 
**Function**
I have added theses function: 
- `getLatestPSP()`: to get the latest PSPs  from a file. 
- `GetPSPbyNonceFromFile()`: A function that will return the PSPs desired for a nonce given. 
- `getNonceSafe()`: A function to return the safe onchain. 
- `ReturnCorrectChainID()`: 
 **Structure**: 
I have added the struct `PSP` that is structure that represent a PSP JSON. 
Also, made modification to the `Defender` to change the order of the arguments in the correct order.

I also improve the overall quality of the code to make sure error are treated as expected. Changing some better type for the address to `common.address`. 

**Tests**
I have improve and add new test that can be saw in the current CI: 
- `TestIsValidHex()`: Test the function `isValidHexString()`.
- `TestGetPSPbyNonceFromFile()`: Test function `GetPSPbyNonceFromFile()`.
- `TestGetNonceSafe()`: Test the function `GetNonceSafe()`.
- `TestReturnCorrectChainID()`: Test the function `ReturnCorrectChainID()`.

**Additional context**

I can help review as the PR is large. 

**Metadata**

- Fixes https://github.com/ethereum-optimism/security-pod/issues/148

### CODE REVIEW

File: op-defender/psp_executor/README.md [hunk 1]
```diff
@@ -8,28 +8,75 @@ The service is designed to listen on a port and execute a PSP onchain when a req
...
 
-To use the HTTP API you can use the following `curl` command with the following fields:
+![cde1a10c6f3eada39b2c27775faefd093623471db95bde4afcf06b9d43f3211b](https://github.com/user-attachments/assets/a75df38c-df06-4195-831b-08edb036a343)
 
-![image](https://github.com/user-attachments/assets/3edc2ee5-6dfd-4872-9bc6-e3ead7444a96)
+To use the HTTP API you can use the following `curl` command with the following fields:
 
 ```bash
-curl -X POST http://${HTTP_API_PSP}:${PORT}/api/psp_execution \-H "Content-Type: application/json" \-d '{"pause": true, "timestamp": 1719432011, "operator": "Tom"}'
+curl -X POST http://localhost:8080/api/psp_execution \-H "Content-Type: application/json" \-d '{"Pause":true,"Timestamp":1596240000,"Operator":"tom"}'
```
Comment (raffaele-oplabs) [inline op-defender/psp_executor/README.md:79]: Should Timestamp and Operator be set by the server?

File: op-defender/psp_executor/README.md [hunk 2]
```diff
@@ -8,28 +8,75 @@ The service is designed to listen on a port and execute a PSP onchain when a req
...
+| `--path` | /tmp/psps.json | Path to JSON file containing PSPs |
+| `--chainid` | 11155111 | Chain ID for the network |
+| `--superchainconfig.address` | 0xC2Be75506d5724086DEB7245bd260Cc9753911Be | Address of SuperchainConfig contract |
+| `--rpc.url` | http://localhost:8545 | URL of the RPC node |
+| `--port.api` | 8080 | Port for the HTTP API server |
+
+**PSPs Format**
+The PSPs are stored with a JSON format. The JSON file should contain an array of PSPs. Each PSP should have the following fields:
+
+```JSON
```
Comment (raffaele-oplabs) [inline op-defender/psp_executor/README.md:43]: Maybe yaml would be better for configuration for readability and the option to add comments?
Not a strong opinion, just adding as possibility.

File: op-defender/psp_executor/README.md [hunk 3]
```diff
@@ -45,27 +92,29 @@ To monitor the _PSPexecutor service_ the metrics server can be enabled. The metr
...
-   --log.level value           The lowest log level that will be output (default: INFO) [$MONITORISM_LOG_LEVEL]
-   --log.format value          Format the log output. Supported formats: 'text', 'terminal', 'logfmt', 'json', 'json-pretty', (default: text) [$MONITORISM_LOG_FORMAT]
-   --log.color                 Color the log output if in terminal mode (default: false) [$MONITORISM_LOG_COLOR]
-   --metrics.enabled           Enable the metrics server (default: false) [$MONITORISM_METRICS_ENABLED]
-   --metrics.addr value        Metrics listening address (default: "0.0.0.0") [$MONITORISM_METRICS_ADDR]
-   --metrics.port value        Metrics listening port (default: 7300) [$MONITORISM_METRICS_PORT]
-   --loop.interval.msec value  Loop interval of the monitor in milliseconds (default: 60000) [$MONITORISM_LOOP_INTERVAL_MSEC]
-   --help, -h                  show help
+   --rpc.url value                   Node URL of a peer (default: "http://127.0.0.1:8545") [$PSPEXECUTOR_NODE_URL]
+   --privatekey value                Privatekey of the account that will issue the pause transaction [$PSPEXECUTOR_PRIVATE_KEY]
```
Comment (raffaele-oplabs) [inline op-defender/psp_executor/README.md:106]: Maybe would be better to have the option to have the private key also in a file as well?

File: op-defender/psp_executor/api_test.go [hunk 4]
```diff
@@ -214,32 +233,35 @@ func TestCheckAndReturnRPC(t *testing.T) {
...
-		{"Empty string", "", "", true},
-		{"Invalid hex string", "0xInvalidHex", "", true},
-		{"Incorrect length", "0x1234", "", true},
-		{"Invalid private key", "0x0000000000000000000000000000000000000000000000000000000000000000", "", true},
+		{"Valid private key", "0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef", hardCodedTestPrivateKey, false},
+		{"Valid private key without 0x", "1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef", hardCodedTestPrivateKey, false},
+		{"Valid private key Generated", validPrivateKeyGeneratedStr, validPrivateKeyGenerated, false},
+		{"Empty string", "", nil, true},
+		{"Invalid hex string", "0xInvalidHex", nil, true},
+		{"Incorrect length", "0x1234", nil, true},
```
Comment (raffaele-oplabs) [inline op-defender/psp_executor/api_test.go:263]:
```suggestion
		{"Incorrect length", "0x1234", nil, true},
		{"Incorrect length", "0x1234123412341234123412341234123412341234123412341234123412341234123412341234", nil, true},

```

**Review by raffaele-oplabs — CHANGES_REQUESTED** [review]
There are some parameters send from command line like user and timestamp, that should be set by the server side to avoid tampering.

File: op-defender/psp_executor/README.md (hunk 2 above)
Comment (Ethnical) [reply op-defender/psp_executor/README.md:43]: Thanks for the suggestion!
 Actually this is the output we have from superchains-ops (this is JSON).
-> https://github.com/ethereum-optimism/superchain-ops/tree/main/tasks/sep/017-presigned-pause
For me, this is not good to reparse to YAML from to JSON each time (if the fields are changing etc).

File: op-defender/psp_executor/README.md (hunk 3 above)
Comment (Ethnical) [reply op-defender/psp_executor/README.md:106]: Interesting, for the current specs the idea was to fetch the secret from outside.
but agreed for standardization in future PR during improvements.

File: op-defender/psp_executor/README.md (hunk 1 above)
Comment (Ethnical) [reply op-defender/psp_executor/README.md:79]: That's correct, and yes this is part of our current specs (for the analyzer and security oncall) authentication. 
No impact for us and this is currently desired.

File: op-defender/psp_executor/cli.go [hunk 5]
```diff
@@ -12,42 +11,34 @@ const (
 	NodeURLFlagName                 = "rpc.url"
 	PrivateKeyFlagName              = "privatekey"
 	PortAPIFlagName                 = "port.api"
-	ReceiverAddressFlagName         = "receiver.address"
 	DataFlagName                    = "data"
```
Comment (raffaele-oplabs) [inline op-defender/psp_executor/cli.go:14]: Are we using this one?

**Review by raffaele-oplabs — COMMENTED** [review]
there are two files named defender.go
one of these is 
op-defender/psp_executor/defender.go
the other
op-defender/defender.go
is this intended?

File: op-defender/psp_executor/defender.go [hunk 6]
```diff
@@ -134,66 +171,134 @@ func (d *Defender) handlePost(w http.ResponseWriter, r *http.Request) {
...
+		d.log.Error("failed to get the PSPs from a file", "error", err)
+		return err
+	}
+	if operationSafe != d.safeAddress {
+		d.log.Error("the safe address in the file is not the same as the one in the configuration!")
+		return err
 	}
-	PspExecutionOnChain(ctx, d.l1Client, d.superchainconfig, d.privatekey, safeAddress, data)
+	// When all the data is fetched correctly then execute the PSP onchain with the PSP data through the `ExecutePSPOnchain()` function.
+	d.ExecutePSPOnchain(ctx, operationSafe, data)
```
Comment (raffaele-oplabs) [inline op-defender/psp_executor/defender.go:303]: Not checking the return error

File: op-defender/psp_executor/defender.go [hunk 7]
```diff
@@ -134,66 +171,134 @@ func (d *Defender) handlePost(w http.ResponseWriter, r *http.Request) {
...
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
+	response := Response{
+		Message: "PSP executed successfully",
+		Status:  http.StatusOK,
+	}
+	json.NewEncoder(w).Encode(response)
 	return
 }
+func (e *DefenderExecutor) ReturnCorrectChainID(l1client *ethclient.Client, chainID uint64) (*big.Int, error) {
```
Comment (raffaele-oplabs) [inline op-defender/psp_executor/defender.go:190]: Comment on the function missing

File: op-defender/psp_executor/defender.go [hunk 8]
```diff
@@ -291,75 +456,60 @@ func (d *Defender) Close(_ context.Context) error {
...
-		return "", fmt.Errorf("Failed to send transaction: %v", err)
+		return "", fmt.Errorf("failed to send transaction: %v", err)
 	}
 
 	return signedTx.Hash().Hex(), nil
 }
 
 // checkPauseStatus(): Is a function made for checking the pause status of the SuperChainConfigAddress
-func checkPauseStatus(ctx context.Context, l1client *ethclient.Client, SuperChainConfigAddress string) (bool, error) {
+func (d *Defender) checkPauseStatus(ctx context.Context, l1client *ethclient.Client) (bool, error) {
```
Comment (raffaele-oplabs) [inline op-defender/psp_executor/defender.go]: l1client not used

**Review by raffaele-oplabs — CHANGES_REQUESTED** [review]
There are some changes to be applied

**Review by raffaele-oplabs — APPROVED** [review]
LGTM!

**Pull request merged**